core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -558,97 +558,134 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
-        throw new FileAlreadyExistsException(errorMessage);
+    Inode<?> lastInode = null;
+    while (lastInode == null) {
+      // Try to lock the last inode with the lock mode of the path.
+      switch (extensibleInodePath.getLockMode()) {
+        case READ:
+          lastInode = currentInodeDirectory.getChildReadLock(name, lockList);
+          break;
+        case WRITE_PARENT:
+        case WRITE:
+          lastInode = currentInodeDirectory.getChildWriteLock(name, lockList);
+          break;
+        default:
+          // This should not be reachable.
+          LOG.warn("Unexpected lock mode encountered: {}",
+              extensibleInodePath.getLockMode().toString());
       }
-    } else {
-      if (options instanceof CreateDirectoryOptions) {
-        CreateDirectoryOptions directoryOptions = (CreateDirectoryOptions) options;
-        lastInode = InodeDirectory.create(mDirectoryIdGenerator.getNewDirectoryId(),
-            currentInodeDirectory.getId(), name, directoryOptions);
-        // Lock the created inode before subsequent operations, and add it to the lock group.
-        lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
-        if (directoryOptions.isPersisted()) {
-          toPersistDirectories.add(lastInode);
+      if (lastInode != null) {
+        // inode to create already exists
+        if (lastInode.isDirectory() && options instanceof CreateDirectoryOptions && !lastInode
+            .isPersisted() && options.isPersisted()) {
+          // The final path component already exists and is not persisted, so it should be added
+          // to the non-persisted Inodes of traversalResult.
+          InodeUtils
+              .syncPersistDirectory((InodeDirectory) lastInode, this, mMountTable, journalContext);
+        } else if (!lastInode.isDirectory() || !(options instanceof CreateDirectoryOptions
+            && ((CreateDirectoryOptions) options).isAllowExists())) {
+          String errorMessage = ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path);
+          LOG.error(errorMessage);
+          throw new FileAlreadyExistsException(errorMessage);
         }
-        lastInode.setPinned(currentInodeDirectory.isPinned());
-      } else if (options instanceof CreateFileOptions) {
-        CreateFileOptions fileOptions = (CreateFileOptions) options;
-        lastInode = InodeFile.create(mContainerIdGenerator.getNewContainerId(),
-            currentInodeDirectory.getId(), name, System.currentTimeMillis(), fileOptions);
-        // Lock the created inode before subsequent operations, and add it to the lock group.
-        lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
-        if (currentInodeDirectory.isPinned()) {
-          // Update set of pinned file ids.
-          mPinnedInodeFileIds.add(lastInode.getId());
+      } else {
+        // create the new inode, with a write lock
+        if (options instanceof CreateDirectoryOptions) {
```

Comment:
Not related to this PR, but is the code overlap large enough to justify switching on creating a file or directory based on the options type instead of having separate methods?